### PR TITLE
Add semantic segmentation example script, no trainer

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds an example script regarding fine-tuning any model supported by the `AutoModelForSemanticSegmentation` API on a semantic segmentation dataset, including regularly pushing to the hub during training as well as WandB logging.

I switched to using Accelerate as I had a bug with the Trainer. 

To do:

- [ ] fix Tensorboard logs

### CODE REVIEW

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
+    else:
+        args.num_train_epochs = math.ceil(args.max_train_steps / num_update_steps_per_epoch)
+
+    lr_scheduler = get_scheduler(
+        name=args.lr_scheduler_type,
+        optimizer=optimizer,
+        num_warmup_steps=args.num_warmup_steps,
+        num_training_steps=args.max_train_steps,
+    )
```
Comment (muellerzr) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:472]: The lr scheduler now gets prepared by `Accelerate` as well
```suggestion
    # Scheduler and math around the number of training steps.
    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
    if args.max_train_steps is None:
        args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
    else:
        args.num_train_epochs = math.ceil(args.max_train_steps / num_update_steps_per_epoch)

    lr_scheduler = get_scheduler(
        name=args.lr_scheduler_type,
        optimizer=optimizer,
        num_warmup_steps=args.num_warmup_steps,
        num_training_steps=args.max_train_steps,
    )
    
    # Prepare everything with our `accelerator`.
    model, optimizer, train_dataloader, eval_dataloader, lr_scheduler = accelerator.prepare(
        model, optimizer, train_dataloader, eval_dataloader, lr_scheduler
    )
```

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+                for k, v in train_logs.items():
+                    if isinstance(v, torch.Tensor):
+                        log_str += "| {}: {:.3e}".format(k, v.item())
+                    else:
+                        log_str += "| {}: {:.3e}".format(k, v)
+
+                if accelerator.is_local_main_process:
+                    progress_bar.write(log_str)
+                    if is_wandb_available() and args.wandb:
+                        wandb.log(train_logs)
```
Comment (muellerzr) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:550]: This can be refactored some based on your preference. I recommend something like this:
```suggestion
                if accelerator.is_local_main_process:
                    progress_bar.write(log_str)
                if args.wandb:
                    accelerator.log(train_logs)
```

**Review by muellerzr — CHANGES_REQUESTED** [review]
Thanks @NielsRogge!

I added a few suggestions based on the recent upgrades to Accelerate and the new script changes that got pushed. 

The other two capabilities we're trying to include in the scripts are [resuming from a checkpoint](https://github.com/huggingface/transformers/blob/main/examples/pytorch/multiple-choice/run_swag_no_trainer.py#L500) as well as `save_state` by [batch](https://github.com/huggingface/transformers/blob/main/examples/pytorch/multiple-choice/run_swag_no_trainer.py#L541) or by [epoch](https://github.com/huggingface/transformers/blob/main/examples/pytorch/question-answering/run_qa_no_trainer.py#L882) (this is just `torch.save`, so slightly different than transformers save. Pick your poison on those I think.)

Nice start!

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+            import wandb
+
+            wandb.init(project=args.output_dir.split("/")[-1])
+    else:
+        datasets.utils.logging.set_verbosity_error()
+        transformers.utils.logging.set_verbosity_error()
+
+    # If passed along, set the training seed now.
+    if args.seed is not None:
+        set_seed(args.seed)
```
[the remark this answers is not in the document]
Comment (sgugger) [reply examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:325]: They were written at the time the one in Accelerate was the same. You should definitely use the one from accelerate with  `device_specific=True` since you have some data augmentation here (otherwise all GPUs will do the exact same data augmentation).

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+
+    # Load pretrained model and feature extractor
+    config = AutoConfig.from_pretrained(
+        args.model_name_or_path, num_labels=num_labels, id2label=id2label, label2id=label2id
+    )
+    feature_extractor = AutoFeatureExtractor.from_pretrained(args.model_name_or_path)
+    model = AutoModelForSemanticSegmentation.from_pretrained(
+        args.model_name_or_path,
+        config=config,
+    )
```
Comment (sgugger) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:380]:
```suggestion
    model = AutoModelForSemanticSegmentation.from_pretrained(args.model_name_or_path, config=config)
```

**Review by sgugger — APPROVED** [review]
Very nice addition! Agreed with all the changes Zach suggested as Accelerate can now handle the logging for you, and the scheduler should be passed along to `prepare` for better support with DeepSpeed.

This will make your example rely on Accelerate master for a couple of weeks until next release (probably next week) so you should add a warning in the Readme to install Accelerate from source. I'll put a not to adapt this after Accelerate next release.

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (muellerzr) [reply examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:325]: Hmm... that might be something that needs adjusting then, as the Accelerate one is different 😄 

cc @sgugger

**Comment by NielsRogge** [comment]
I've added all requested changes. As you can see, I'm updating the repo using commits rather than creating subfolders.

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+                completed_steps += 1
+
+            if isinstance(checkpointing_steps, int):
+                if completed_steps % checkpointing_steps == 0:
+                    accelerator.save_state(args.output_dir)
+
+                    # Optionally push to the hub
+                    if args.push_to_hub and accelerator.is_main_process:
+                        feature_extractor.save_pretrained(args.output_dir)
+                        repo.push_to_hub(
```
Comment (sgugger) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:542]: Note that the checkpoint saved here is saved with `torch.save` and not `save_pretrained` so they're not usable with `from_pretrained`.

Also, you'll need to adapt the logic to resume from a checkpoint to checkout the right branch in the folder if you go this way, as everything is saved in the root of `output_dir`.

Comment (NielsRogge) [reply examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py]: I've removed this logic and updated the script to be aligned with the other example scripts.

**Comment by NielsRogge** [comment]
Some issue with git, closing this issue in favor of a new one.
